flocker/node/agents/cinder.py

```diff
@@ -852,17 +854,53 @@ def get_keystone_session(**config):
         )
 
 
-def get_cinder_v1_client(session, region):
+SUPPORTED_VERSIONS = (
+    (u"2", u"volumev2"),
+    (u"1", u"volume"),
+)
+
+
+class CinderAPIVersionDetectionFailure(Exception):
+    """
+    Unable to connect to a supported version of the Cinder API.
+    """
+    _template = "CinderAPIVersionDetectionFailure(endpoint_errors={!r})"
+
+    def __init__(self, endpoint_errors):
+        self.endpoint_errors = endpoint_errors
+
+    def __str__(self):
+        return self._template.format(
+            self.endpoint_errors,
+        )
+
+    __repr__ = __str__
+
+
+def get_cinder_client(session, region):
     """
     Create a Cinder (volume) client from a Keystone session.
 
+    Try all ``SUPPORTED_VERSIONS`` in order and return the first client to
+    successfully complete a ``list`` API call.
+
     :param keystoneclient.Session session: Authenticated Keystone session.
     :param str region: Openstack region.
-    :return: A cinderclient.Client
+    :return: A cinderclient.Client (v2 else v1)
     """
-    return CinderClient(
-        session=session, region_name=region, version=1
-    )
+    endpoint_errors = []
+    for version, service_type in SUPPORTED_VERSIONS:
```

Comment:
I missed that. It was left over from a previous implementation where I attempted to lookup the available Cinder API services in the OpenStack Keystone service catalog. 
Unfortunately, Rackspace deny access to the service_catalog endpoint.

I've removed the ``service_type``.